### PR TITLE
Add a horovod optimizer and example

### LINKED ISSUE #2570: FR distributed training via Horovod
What would it take to support multiple GPUs in SVI, as [requested on the forum](https://forum.pyro.ai/t/run-pyro-on-multiple-gpus/1066)?

See Horovod's [pytorch](https://github.com/horovod/horovod/blob/master/docs/pytorch.rst) [examples](https://github.com/horovod/horovod/blob/master/examples/pytorch_mnist.py).

## Tentative Tasks
- [x] Support some combination of `hvd.DistributedOptimizer` with `PyroOptimizer`. This probably requires extra initialization step to discover all parameters before inference. This definitely requires a static parameter set.
- [x] Add any required initialization logic to SVI
- [x] Add an example or tutorial

cc @mbabadi @dalessioluca

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #2570 

This integrates with [Horovod](https://github.com/horovod/horovod) by implementing a `pyro.optim.HorovodOptimizer` that combines behavior of Pyro's `PyroOptim` and Horovod's `DistributedOptimizer`. The new `HorovodOptimizer` can wrap any other Pyro optimizer e.g. `optim = HorovodOptimizer(ClippedAdam({'lr': 1e-3}))`. I've also added an example demonstrating how to distribute training using the optimizer.

I have avoided adding horovod as a default dependency since it requires other system dependencies (MPI and maybe gloo). Instead I've added a `pyro[horovod]` extras require which points to `horovod[pytorch]`. I've also added a test decorator `requires_horovod` that skips on current CI.

## Tested
Tests with `--no-horovod` are run on CI. Other tests require `horovodrun` and are not run on CI. Because horovod must be run under a separate process, I have added only example tests and no pytest-driven unit tests.

- [x] smoke test `--no-horovod`
- [x] smoke test `--num-proc 1`
- [x] smoke test `--num-proc 2`
- [x] smoke test `--no-horovod` `--cuda`
- [x] smoke test `--num-proc 1` `--cuda`
- [x] smoke test `--num-proc 2` `--cuda`
- [x] verify convergence

### CODE REVIEW

**Comment by fritzo** [comment]
@fehiepsi @martinjankowiak do either of you want to review this?

File: examples/svi_horovod.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        coeff = pyro.sample("coeff", dist.Normal(0, 1))
+        bias = pyro.sample("bias", dist.Normal(0, 1))
+        scale = pyro.sample("scale", dist.LogNormal(0, 1))
+        with pyro.plate("data", self.size, len(covariates)):
+            loc = bias + coeff * covariates
+            return pyro.sample("obs", dist.Normal(loc, scale),
+                               obs=data)
+
+
+# The following is a standard training loop. To epmhasize the Horovod-specific
```
Comment (martinjankowiak) [inline examples/svi_horovod.py:57]: typo: epmhasize

File: pyro/optim/horovod.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    r"""
+    Distributed wrapper for a :class:`~pyro.optim.optim.PyroOptim` optimizer.
+
+    This class wraps a ``PyroOptim`` object similar to the way
+    :func:`horovod.torch.DistributedOptimizer` wraps a
+    :class:`torch.optim.Optimizer`.
+
+    .. note::
+
+        This requires :mod:`horovod.torch` is installed, e.g. via
```
Comment (martinjankowiak) [inline pyro/optim/horovod.py:19]: grammar typo

File: examples/svi_horovod.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+class Model(PyroModule):
+    def __init__(self, size):
+        super().__init__()
+        self.size = size
+
+    def forward(self, covariates, data=None):
+        coeff = pyro.sample("coeff", dist.Normal(0, 1))
+        bias = pyro.sample("bias", dist.Normal(0, 1))
+        scale = pyro.sample("scale", dist.LogNormal(0, 1))
+        with pyro.plate("data", self.size, len(covariates)):
```
Comment (martinjankowiak) [inline examples/svi_horovod.py:51]: so what happens if the user relies on `plate`-controlled subsampling? the same computation happens on all workers?

Comment (fritzo) [reply examples/svi_horovod.py:57]: Good point, I've added a couple sentences explaining requirements.

**Comment by fritzo** [comment]
Thanks for reviewing!

**Pull request merged**